radicle/src/identity/project.rs

```diff
@@ -20,42 +25,168 @@ pub enum ProjectError {
+    ///
+    /// # Validation Rules
+    ///
+    ///   * `name`'s length must not be empty and must not exceed 255.
+    ///   * `description`'s length must not exceed 255.
+    ///   * `default_branch`'s length must not be empty and must not exceed 255.
+    pub fn new(
+        name: String,
+        description: String,
+        default_branch: BranchName,
+    ) -> Result<Self, Vec<ProjectError>> {
+        let mut errs = Vec::new();
+
+        if name.is_empty() {
+            errs.push(ProjectError::Name("name cannot be empty"));
+        } else if name.len() > doc::MAX_STRING_LENGTH {
+            errs.push(ProjectError::Name("name cannot exceed 255 bytes"));
         }
-        if self.description.len() > doc::MAX_STRING_LENGTH {
-            return Err(ProjectError::Description(
+
+        if description.len() > doc::MAX_STRING_LENGTH {
+            errs.push(ProjectError::Description(
                 "description cannot exceed 255 bytes",
             ));
         }
-        if self.default_branch.is_empty() {
-            return Err(ProjectError::DefaultBranch(
+
+        if default_branch.is_empty() {
+            errs.push(ProjectError::DefaultBranch(
                 "default branch cannot be empty",
-            ));
-        }
-        if self.default_branch.len() > doc::MAX_STRING_LENGTH {
-            return Err(ProjectError::DefaultBranch(
+            ))
+        } else if default_branch.len() > doc::MAX_STRING_LENGTH {
+            errs.push(ProjectError::DefaultBranch(
                 "default branch cannot exceed 255 bytes",
-            ));
+            ))
         }
-        Ok(())
+
+        if errs.is_empty() {
+            Ok(Self {
+                name,
+                description,
+                default_branch,
+            })
+        } else {
+            Err(errs)
+        }
+    }
+
+    /// Update the `Project` payload with new values, if provided.
+    ///
+    /// When any of the values are set to `None` then the original
+    /// value will be used, and so the value will pass validation.
+    ///
+    /// Otherwise, the new value is used and will be subject to the
+    /// original validation rules (see [`Project::new`]).
+    pub fn update(
+        self,
+        name: impl Into<Option<String>>,
+        description: impl Into<Option<String>>,
+        default_branch: impl Into<Option<BranchName>>,
+    ) -> Result<Self, Vec<ProjectError>> {
+        let name = name.into().unwrap_or(self.name);
+        let description = description.into().unwrap_or(self.description);
+        let default_branch = default_branch.into().unwrap_or(self.default_branch);
+        Self::new(name, description, default_branch)
+    }
+
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+
+    pub fn description(&self) -> &String {
+        &self.description
+    }
+
+    pub fn default_branch(&self) -> &BranchName {
+        &self.default_branch
     }
```

Comment:
Would inlining remove the overhead of the function call?  The methods seem added purely for safeguarding against changes.

```
#[inline]
pub fn default_branch(&self) -> &BranchName {
    &self.default_branch
}
```

Reply:
Ya, that makes sense to me!